### PR TITLE
Deprecated hook handling

### LINKED ISSUE #13132: [2.7] No Backward Compatibility for Many Deprecated Hooks
## EXPLANATION OF THE ISSUE

There are many deprecated hooks in WC 2.7 that will no longer be triggered, breaking backward compatibility with any code that has not been updated to use new 2.7 hooks.

Any hooks called in a deprecated function are affected, because the deprecated function still needs to be called in order for the hook to be triggered. Core doesn't call these functions anymore, nor will 3rd party code which has been updated for 2.7, meaning the hooks won't be triggered.

For example:

1. the `WC_Abstract_Order::add_shipping()` method has been deprecated in favour of using `WC_Abstract_Order::add_item()`
2. `add_shipping()` triggers the `'woocommerce_order_add_shipping'` hook
3. Although `WC_Abstract_Legacy_Order::add_shipping()` will make sure `'woocommerce_order_add_shipping'` is still triggered, it will only do so when `add_shipping()` is actually called on an instance of `WC_Order`
4. Neither core nor updated 3rd party code will ever call `add_shipping()`
5. The `'woocommerce_order_add_shipping'` hook will never be triggered
6. Any 3rd party code attached to the `'woocommerce_order_add_shipping'` hook will never run
7. The deprecated hook notice also won't be displayed, meaning the 3rd party code won't even know it's using the wrong hook. It will just silently fail.

That's just one of many examples of broken hooks.

Not triggering these hooks has the potential to break a huge amount of functionality from 3rd party code. Silently failing will make diagnosing the cause very difficult, and time consuming, for support ninjas and developers.

#### EXAMPLE SOLUTION

We faced a similar issue when we deprecated a lot of hooks with Subscriptions 2.0. We mapped old hooks to new ones via a base [hook deprecator class](https://github.com/Prospress/woocommerce-subscriptions/blob/master/includes/abstracts/abstract-wcs-hook-deprecator.php) that was extended for different types of hooks, like [filters](https://github.com/Prospress/woocommerce-subscriptions/blob/master/includes/deprecated/class-wcs-filter-deprecator.php) and even [actions with variable names](https://github.com/Prospress/woocommerce-subscriptions/blob/master/includes/deprecated/class-wcs-dynamic-action-deprecator.php). That way, the old functions didn't need to be called for the old hook to also be called and a deprecated notice to be logged or displayed.

Manually mapping all the old hooks to new ones takes a lot of work, but it saves breaking stores running code that hadn't been updated. The number of stores running WooCommerce is an order of magnitude higher than Subscriptions, making the impact of this issue in WC core far worse than it was in Subscriptions, and the work required to address it far more worthwhile.

## STEPS TO REPRODUCE THE ISSUE

On a store running WooCommerce < 2.7:

1. Copy & paste the snippet below somewhere to run on the site
1. Go to **WooCommerce > Edit Order**
1. Click **Add Item(s)**
1. Click **Add Shipping Cost**
1. Notice that an entry is added to the error log

Repeat these steps on a site with WC 2.7 active. No entry is added to the error log, because `'woocommerce_order_add_shipping` is never triggered.

Test snippet: 

```
function eg_log_shipping_hook( $order_id, $item_id, $shipping_rate ) {
	error_log( 'In ' . __METHOD__ . '() for order ' . $order_id . ' with item ' . $item_id . ' $shipping_rate = ' . print_r( $shipping_rate, true ) );
}
add_action( 'woocommerce_order_add_shipping', 'eg_log_shipping_hook', 10, 3 );
```

### PR DESCRIPTION
**Pull request by mikejolley**
Closes #13132 

Adds deprecation handling classes based on @thenbrent’s work to deprecate actions and filters.

Also adds correct deprecation methods for soft deprecated functions.

Maps the order line item actions to the replacements - `$item` is already backwards compatible so can be passed to the hooks directly without modification.

### CODE REVIEW

File: includes/wc-deprecated-functions.php [hunk 1]
```diff
@@ -14,6 +14,12 @@
 	exit;
 }
 
+include_once( 'abstracts/abstract-wc-deprecated-hooks.php' );
+include_once( 'class-wc-deprecated-action-hooks.php' );
```
Comment (claudiulodro) [inline includes/wc-deprecated-functions.php]: Should there also be a `include_once( 'class-wc-deprecated-filter-hooks.php' );` here?

**Review by claudiulodro — CHANGES_REQUESTED** [review]

File: includes/wc-deprecated-functions.php (hunk 1 above)
Comment (claudiulodro) [inline includes/wc-deprecated-functions.php]: Other than that, the logic of the deprecated filter classes looks solid to me. You can make a new issue if you want me to write some tests for these classes to verify.

Comment (mikejolley) [inline includes/wc-deprecated-functions.php]: Thanks 👍 I missed that.

**Comment by mikejolley** [comment]
@claudiulodro How would you approach tests for this? Firing the actions/filters manually? Would be super interested to see if you wanted to give it a shot.

**Comment by thenbrent** [comment]
So glad to see all those deprecated hooks being triggered now @mikejolley! 😂 

> How would you approach tests for this?

We tested it manually with Subscriptions 2.0. It probably won't be too much help, but this is the plugin we used: https://gist.github.com/thenbrent/53cfdbff0533c85e89e5

**Pull request merged**